cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/decoder/FormatStreamEventDecoder.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright © 2014 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data.stream.decoder;
+
+import co.cask.cdap.api.data.format.RecordFormat;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.flow.flowlet.StreamEvent;
+import co.cask.cdap.api.stream.StreamEventDecoder;
+import com.google.common.collect.Lists;
+import org.apache.hadoop.io.LongWritable;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+
+/**
+ * A {@link StreamEventDecoder} that decodes {@link StreamEvent} into {@link LongWritable} as key
+ * and {@link StructuredRecord} as value for Mapper input. The key carries the event timestamp, while
+ * the value is the stream event body formatted by some {@link RecordFormat}.
+ */
+public final class FormatStreamEventDecoder implements StreamEventDecoder<LongWritable, StructuredRecord> {
+  private static final String TIMESTAMP_FIELD = "ts";
+  private static final String HEADERS_FIELD = "headers";
+  private final LongWritable key = new LongWritable();
+  private final RecordFormat<ByteBuffer, StructuredRecord> bodyFormat;
+  private final Schema eventSchema;
+  private final Schema bodySchema;
+
+  /**
+   * Create a decoder for stream events that decodes the body of the stream using the given initialized format.
+   *
+   * @param bodyFormat Initialized format.
+   */
+  public FormatStreamEventDecoder(RecordFormat<ByteBuffer, StructuredRecord> bodyFormat) {
+    this.bodyFormat = bodyFormat;
+    this.bodySchema = bodyFormat.getSchema();
+    List<Schema.Field> fields = Lists.newArrayList(
+      Schema.Field.of(TIMESTAMP_FIELD, Schema.of(Schema.Type.LONG)),
```

Comment:
Why need to repeat the timestamp in the record if it is already in the mapper key?